File: python_prototype/CoordsManager.py
import math


def equirectangular_approximation(lat1, long1, lat2, long2):
    EARTH_RADIUS = 6371
    delta_longitude = math.radians(long2 - long1)
    delta_latitude = math.radians(lat2 - lat1)
    sum_latitude = math.radians(lat1 + lat2)

    x = delta_longitude * math.cos(sum_latitude / 2.0)
    distance = EARTH_RADIUS * math.sqrt(x * x + delta_latitude * delta_latitude)
    return distance


class CoordsManager:
    def __init__(self, observer_location, coords_range, grid_size):
        self.latitude_range = coords_range[0]
        self.longitude_range = coords_range[1]
        self.grid_size = grid_size
        self.observer_location_geo = observer_location

    def convert_geo_to_local_coords(self, latitude, longitude, altitude):
        z = (longitude - self.longitude_range[0]) * self.grid_size[1]
        x = (self.latitude_range[1] - latitude) * self.grid_size[0]
        altitude_scale = 1 / 1000  # altitude_scale is 0.001 kilometer (1 meter => 1 value in elevation data == 1 meter)
        y = altitude * altitude_scale

        return x, y, z

    def calc_distance_observer_to_point(self, latitude, longitude):
        return equirectangular_approximation(latitude, longitude,
                                             self.observer_location_geo[0], self.observer_location_geo[1])

File: app/src/main/java/com/example/peaksrecognition/CoordsManager.java
package com.example.peaksrecognition;

public class CoordsManager {
    private static final double EARTH_RADIUS = 6371.0;
    private final double[] latitudeRange;
    private final double[] longitudeRange;
    private final double altitudeScale = 1.0 / 1000.0;
    private final double[] gridSize;
    private final double[] observerLocationGeo;

    CoordsManager(double[] observerLocationGeo, double[][] coordsRange, double[] gridSize) {
        latitudeRange = coordsRange[0];
        longitudeRange = coordsRange[1];
        this.gridSize = gridSize;
        this.observerLocationGeo = observerLocationGeo;
    }

    public static double equirectangularApproximation(double latitude1, double longitude1, double latitude2, double longitude2) {
        double deltaLongitude = Math.toRadians(longitude2 - longitude1);
        double deltaLatitude = Math.toRadians(latitude2 - latitude1);
        double sumLatitude = Math.toRadians(latitude1 + latitude2);

        double x = deltaLongitude * Math.cos(sumLatitude / 2.0);
        return EARTH_RADIUS * Math.sqrt(x * x + deltaLatitude * deltaLatitude);
    }

    public double[] convertGeoToLocalCoords(double latitude, double longitude, double altitude) {
        double x = (latitudeRange[1] - latitude) * gridSize[0];
        double z = (longitude - longitudeRange[0]) * gridSize[1];
        double y = altitude * altitudeScale;

        return new double[]{x, y, z};
    }

    public double calcDistanceObserverToPoint(double latitude, double longitude) {
        return equirectangularApproximation(latitude, longitude,
                observerLocationGeo[0], observerLocationGeo[1]);
    }
}
